### ISSUE TITLE
Does it support Authorisation mode IAM? i.e. IAM_AUTH

### ISSUE THREAD

**Issue by pyros2097** [sony/appsync-client-go#1]
I'm trying to make a request in one of my lambda's which has IAM permission to access my graphql api. I don't see how to do it in the examples.

**Comment by pyros2097** [comment]
@leandrogr I figured it out. For AWS_IAM you need to sign the http request with the aws-sdk (basically add the headers the signing part spits out into your request) thats all. If you are running it in a lambda just make sure you have assigned the right permissions for the lambda. If you are running it locally it should just work fine.

**Comment by KwanJunWen** [comment]
@pyros2097 Hello, is it possible for you to provide some code examples on how to sign the request ?

**Comment by pyros2097** [comment]
@KwanJunWen I had done it but left the organisation. On top of my head it was like this,
use this func to sign the headers of your request,

https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/#Signer.Sign

Pass the body also to the func and I think the service is `appsync` and then the request will work only if you have authorisation through AWS_IAM enabled on the API and also added @aws_iam decorator on the type you need to query for in the schema.

**Comment by KwanJunWen** [comment]
In case anyone wondering. Here are some working example codes that can be used as reference
I've used aws v4 signer to sign the request as @pyros2097 mentioned.
https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/#Signer.Sign

```go

// Initializes new session
sess, err := session.NewSession(&aws.Config{
	Region: aws.String("ap-southeast-1"),
})
if err != nil {
	return err
}

// Construct GraphQL request
graphqlReq := graphql.PostRequest{
	Query: `
		query foo {
			foo (
				id: "f-001"
			) {
				id
				bar
			}
		}
	`,
}

// Get GraphQL request blob
qb, err := json.Marshal(&graphqlReq)
if err != nil {
	return err
}

// Initializes new signer with credentials
signer := v4.NewSigner(sess.Config.Credentials)

// Initializes new http request for signer
r, err := http.NewRequest(http.MethodPost, "ENDPOINT", nil)
if err != nil {
	return err
}

// Set Content-Type header to application/json before signing request
r.Header.Set("Content-Type", "application/json")

// Sign http request with GraphQL request body
_, err = signer.Sign(r, bytes.NewReader(qb), "appsync", "ap-southeast-1", time.Now())
if err != nil {
	return nil, err
}

// Initializes new GraphQL client with appsync-client package
graphqlClient := appsync.NewGraphQLClient(graphql.NewClient("ENDPOINT"))

// Post request with v4 signed http request header and GraphQL request
resp, err := graphqlClient.Post(r.Header, graphqlReq)
if err != nil {
	return err
}

data := new(string)
if err := resp.DataAs(data); err != nil {
	return err
}
```
@pyros2097 Thanks for the quick reply !
I've managed to make it work right now. Is there anyway to call the endpoint with
```go
// Initializes new AppSync client with appsync-client package
appsyncClient := appsync.NewClient(appsync.NewGraphQLClient(graphql.NewClient(
	"ENDPOINT",
	graphql.WithCredential(r.Header.Get("Authorization")),
)))
resp, err := appsyncClient.Post(graphqlReq)
if err != nil {
	return err
}
```
Apparently I still get Forbidden error even after setting `graphql.WithCredential`. I need to pass in the signed http request header directly into graphql client instead.

**Comment by pyros2097** [comment]
Yes I didn't use this graphql client. I directly used golang http client and assigned the headers. 
The signer.Sign returns the headers and you need to do r.Header = headers and then it should work.

**Comment by KwanJunWen** [comment]
I see, I was able to call appsync with standard http requests as well. I was trying to use this package with v4 signed request.

Apparently we will have to create a separate http request just to get the request signed and then pass in the signed request header into graphql client to post request in order for it to work. The `graphql.WithCredential` does not work as intended. It will be great if we could v4 sign the graphql request directly so we won't have to setup a separate http request just for the signature.

Yes, perhaps we could open a PR to address this issue.

**Comment by macbutch** [comment]
Thank you @KwanJunWen. Your example is very useful.

**Comment by KazuhitoNakamura** [comment]
Hi @pyros2097 

Sorry for our very late reply and not updating some features.
We have released v1.1.0 for IAM Authorization.
If possible, please refer to this [sample](https://github.com/sony/appsync-client-go/blob/master/test/EventApp/appsync_eventapp.go#L48).

**Comment by pyros2097** [comment]
Nice. I guess this can be closed now.
